src/keys/bip39.rs

```diff
@@ -5,24 +5,132 @@
 
 // https://doc.rust-lang.org/std/primitive.str.html
 // "String slices are always valid UTF-8."
-type Mnemonic = str;
-type Passphrase = str;
-type Seed = [u8; 64];
 
 use alloc::string::{String, ToString};
+use core::convert::TryFrom;
+use core::ops::Deref;
 
-use unicode_normalization::UnicodeNormalization;
+use unicode_normalization::{is_nfkd, UnicodeNormalization};
+use zeroize::{Zeroize, ZeroizeOnDrop, Zeroizing};
 
-pub fn mnemonic_to_seed(m: &Mnemonic, p: &Passphrase, s: &mut Seed) {
-    let m = m.chars().nfkd().collect::<String>();
+#[derive(Debug, PartialEq, Eq)]
+pub enum Error {
+    InvalidEntropyCount(usize),
+    NoSuchWord(String),
+    ChecksumMismatch,
+    UnnormalizedMnemonic,
+    UnnormalizedPassphrase,
+    BadWordlist,
+    BadSeparator,
+}
+
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub struct MnemonicRef<'a>(&'a str);
+
+impl<'a> Deref for MnemonicRef<'a> {
+    type Target = str;
+    fn deref(&self) -> &str {
+        self.0
+    }
+}
+
+impl<'a> TryFrom<&'a str> for MnemonicRef<'a> {
+    type Error = Error;
+    fn try_from(mnemonic_str: &'a str) -> Result<Self, Error> {
+        if is_nfkd(mnemonic_str) {
+            Ok(MnemonicRef(mnemonic_str))
+        } else {
+            Err(Error::UnnormalizedMnemonic)
+        }
+    }
+}
+
+impl<'a> PartialEq<str> for MnemonicRef<'a> {
+    fn eq(&self, other: &str) -> bool {
+        self.0.eq(other)
+    }
+}
+
+#[derive(ZeroizeOnDrop)]
+pub struct Mnemonic(String);
+
+impl From<&str> for Mnemonic {
+    fn from(unnormalized_mnemonic: &str) -> Self {
+        Self(unnormalized_mnemonic.chars().nfkd().collect())
+    }
+}
+
+impl<'a> From<&'a Mnemonic> for MnemonicRef<'a> {
+    fn from(mnemonic_ref: &'a Mnemonic) -> Self {
+        Self(&mnemonic_ref.0)
+    }
+}
+
+impl AsRef<str> for Mnemonic {
+    fn as_ref(&self) -> &str {
+        &self.0
+    }
+}
+
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub struct PassphraseRef<'a>(&'a str);
+
+impl<'a> TryFrom<&'a str> for PassphraseRef<'a> {
+    type Error = Error;
+    fn try_from(passphrase_str: &'a str) -> Result<Self, Error> {
+        if is_nfkd(passphrase_str) {
+            Ok(PassphraseRef(passphrase_str))
+        } else {
+            Err(Error::UnnormalizedPassphrase)
+        }
+    }
+}
 
+#[derive(ZeroizeOnDrop)]
```

Comment:
Can we add Clone here?

Reply:
We sometimes do this, for example here when it's needed in different functions https://github.com/iotaledger/iota-sdk/blob/d1ae85026129ce6c3f439d539dc69a4794c751b3/cli/src/command/wallet.rs#L109-L111

Reply:
I'm not sure about `let password = password.into();`. This just makes another copy of password. Why do you need `impl Into<Password>`? Are there many types convertible into `Password`? Why not just pass passwords by reference? Passwords should not be stored internally (to justify moving value), they should be used to compute keys and encrypt/decrypt (to do that you only need a reference). On the other hand, if a password should be stored then it should be stored in stronghold. The best type for passwords if they are passed by reference is `&[u8]`. In bip39 passwords need to be normalized, hence a separate wrapper.

Reply:
Sorry my example was about another password type we have, not the bip39 one